packages/labs/ssr/src/lib/module-loader.ts

```diff
@@ -0,0 +1,279 @@
+ *  - A module cache keyed by resolved URL
+ *  - import.meta.url support
+ *  - Dynamic import() that functions the same as static imports
+ *
+ * There are some behaviors specific to lit-html. Mainly that imports of certain
+ * directives are redirected to Node/SSR compatible implementations.
+ */
+export class ModuleLoader {
+  private static _nextVmContextId = 0;
+
+  // TODO (justinfagnani): document the need for this
+  private readonly _vmContextId = ModuleLoader._nextVmContextId++;
+
+  private readonly _context: vm.Context;
+
+  /**
+   * TODO (justinfagnani): This is a temporary stand-in for a real graph API.
+   * We want to be able to invalidate a module and the transitive closure
+   * of its importers so that we can update the graph.
+   *
+   * The keys of the map are useful for enumering static imported modules
+   * after an entrypoint is loaded.
+   */
+  readonly cache = new Map<string, ModuleRecord>();
+
+  // TODO (justinfagnani): Allow passing a filesystem object to allow network
+  // sources, in-memory for tests, etc.
+  constructor(contextGlobal: vm.Context) {
+    this._context = vm.createContext(contextGlobal);
+  }
+
+  /**
+   * Imports a module given by `path` into a new VM context with `contextGlobal` as the
+   * global object.
+   *
+   * @param specifier
+   * @param referrer
+   * @param contextGlobal The object that will become the global, via vm.createContext
+   */
+  async importModule(
+    specifier: string,
+    referrer: string
+  ): Promise<ImportResult> {
+    const result = await this._loadModule(specifier, referrer);
+    const {module} = result;
+    if (module.status === 'unlinked') {
+      await result.module.link(this._linker);
+    }
+    if (module.status !== 'evaluated') {
+      await result.module.evaluate();
+    }
+    return result;
+  }
+
+  /**
+   * Performs the actual loading of module source from disk, creates the
+   * Module instance, and maintains the module cache.
+   *
+   * Used directly by `importModule` and by the linker and dynamic import()
+   * support function.
+   */
+  private async _loadModule(
+    specifier: string,
+    referrer: string
+  ): Promise<ImportResult> {
+    if (/:\d+$/.test(referrer)) {
+      throw new Error('used identifier as referrer');
+    }
+
+    if (builtIns.has(specifier)) {
+      return this._loadBuiltInModule(specifier);
+    }
+
+    const moduleURL = resolveSpecifier(specifier, referrer);
+    if (moduleURL.protocol !== 'file:') {
+      throw new Error(`Unsupported protocol: ${moduleURL.protocol}`);
+    }
+    const modulePath = moduleURL.pathname;
+
+    // Look in the cache
+    let moduleRecord = this.cache.get(modulePath);
+    if (moduleRecord !== undefined) {
+      return {
+        path: modulePath,
+        module: await moduleRecord.evaluated,
+      };
+    }
+
+    const modulePromise = (async () => {
+      const source = await fs.readFile(modulePath, 'utf-8');
+      // TODO: store and re-use cachedData:
```

Comment:
Comment still relevant? It is cached now, right? Or if there's another layer of caching we could do, maybe elaborate on the comment (there is nothing called `cachedData`).

Reply:
This is referring to Node's vm module cachedData, which should speed up repeated module creation in cases where we create fresh vm contexts. It's not related to our module cache. I updated the link and it seems like they added cachedData for modules now, so we can do this.